### PR TITLE
Version flags and logging

### PR DESCRIPTION
**Pull request by ashleywang1**
Add --release-version to control what versions of gloo or gloo enterprise to install.

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1863

### CODE REVIEW

File: projects/gloo/cli/pkg/flagutils/install.go [hunk 1]
```diff
@@ -15,11 +15,14 @@ func AddInstallFlags(set *pflag.FlagSet, install *options.Install) {
 	set.StringVar(&install.HelmReleaseName, "release-name", constants.GlooReleaseName, "helm release name")
 	set.BoolVar(&install.CreateNamespace, "create-namespace", true, "Create the namespace to install gloo into")
 	set.StringVarP(&install.Namespace, "namespace", "n", defaults.GlooSystem, "namespace to install gloo into")
+	set.StringVar(&install.Release, "gloo-release", "", "Version of Gloo to install (e.g. 1.0.0, defaults to latest)")
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/flagutils/install.go]:
```suggestion
	set.StringVar(&install.Release, "open-source-release", "", "Version of Gloo to install (e.g. 1.0.0, defaults to latest)")
```
idk, this is the kind of thing that might be good for our new design review meetings, to get consensus on a user-facing change like this

File: projects/gloo/cli/pkg/cmd/install/gateway.go [hunk 2]
```diff
@@ -15,6 +16,9 @@ func gatewayCmd(opts *options.Options) *cobra.Command {
 		PreRun: setVerboseMode(opts),
 
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if opts.Install.Release != "" {
+				version.Version = opts.Install.Release
+			}
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/gateway.go]: This should be reverted. When this line is executed, `version.Version` is already set - to the version that glooctl is. We shouldn't be mutating that value during normal execution

File: projects/gloo/cli/pkg/cmd/install/enterprise.go [hunk 3]
```diff
@@ -21,7 +22,10 @@ func enterpriseCmd(opts *options.Options) *cobra.Command {
 		RunE: func(cmd *cobra.Command, args []string) error {
 
 			extraValues := map[string]interface{}{
-				"license_key": opts.Install.LicenseKey,
+				"license_key": opts.Install.Enterprise.LicenseKey,
+			}
+			if opts.Install.Enterprise.Release != "" {
+				version.EnterpriseTag = opts.Install.Enterprise.Release
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/enterprise.go]: Talking offline, but I suspect we don't want to do this

File: projects/gloo/cli/pkg/flagutils/install.go (hunk 1 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/flagutils/install.go]: Adding a --release-version flag.

File: docs/content/installation/gateway/kubernetes/values.txt [hunk 4]
```diff
@@ -234,8 +234,8 @@
 |accessLogger.resources.requests.cpu|string||amount of CPUs|
 |global.image.tag|string||tag for the container|
 |global.image.repository|string||image name (repository) for the container.|
-|global.image.registry|string|quay.io/solo-io|image prefix/registry e.g. (quay.io/solo-io)|
-|global.image.pullPolicy|string|IfNotPresent|image pull policy for the container|
+|global.image.registry|string|gcr.io/|image prefix/registry e.g. (quay.io/solo-io)|
+|global.image.pullPolicy|string|Always|image pull policy for the container|
```
Comment (GrahamGoudeau) [inline docs/content/installation/gateway/kubernetes/values.txt]: revert?

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 5]
```diff
@@ -188,27 +189,35 @@ func (i *installer) printReleaseManifest(release *release.Release) error {
...
-		enterpriseTag = version.EnterpriseTag
-	}
 
 	if enterprise {
-		helmChartArchiveUri = fmt.Sprintf(GlooEHelmRepoTemplate, enterpriseTag)
+		if releaseVersionOverride != "" {
+			helmChartArchiveUri = fmt.Sprintf(GlooEHelmRepoTemplate, releaseVersionOverride)
+		} else {
+			helmChartArchiveUri = fmt.Sprintf(GlooEHelmRepoTemplate, enterpriseTag)
+		}
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]: can we DRY-out some of this duplication?

File: projects/gloo/cli/pkg/flagutils/install.go [hunk 6]
```diff
@@ -13,8 +13,10 @@ func AddInstallFlags(set *pflag.FlagSet, install *options.Install) {
 	set.StringVarP(&install.HelmChartOverride, "file", "f", "", "Install Gloo from this Helm chart archive file rather than from a release")
 	set.StringSliceVarP(&install.HelmChartValueFileNames, "values", "", []string{}, "List of files with value overrides for the Gloo Helm chart, (e.g. --values file1,file2 or --values file1 --values file2)")
 	set.StringVar(&install.HelmReleaseName, "release-name", constants.GlooReleaseName, "helm release name")
+	set.StringVar(&install.HelmReleaseVersion, "release-version", "", "helm release version (e.g. 0.20.6, defaults to latest)")
 	set.BoolVar(&install.CreateNamespace, "create-namespace", true, "Create the namespace to install gloo into")
 	set.StringVarP(&install.Namespace, "namespace", "n", defaults.GlooSystem, "namespace to install gloo into")
+	set.StringVar(&install.LicenseKey, "license-key", "", "License key to activate GlooE features")
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/flagutils/install.go]: why are we adding this flag? It's still in `AddEnterpriseInstallFlags` below. Can we remove one of them?

File: projects/gloo/cli/pkg/cmd/install/install_test.go [hunk 7]
```diff
@@ -31,12 +31,23 @@ var _ = Describe("Install", func() {
...
 
+	It("shouldn't get errors for enterprise dry run without file", func() {
+		_, err := testutils.GlooctlOut(fmt.Sprintf("install gateway enterprise --dry-run %s", licenseKey))
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("shouldn't get errors when overriding enterprise version", func() {
+		_, err := testutils.GlooctlOut(fmt.Sprintf("install gateway enterprise --file %s --release-version %s --dry-run %s", file, overrideVersion, licenseKey))
+		Expect(err).NotTo(HaveOccurred())
+	})
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/install/install_test.go]: we shouldn't allow providing both `--file` and `--release-version` args. This should error, and we should have a test that it errors

File: projects/gloo/cli/pkg/cmd/install/installer.go (hunk 5 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/install/installer.go]: agreed, can definitely DRY and flatten the code here by doing the `releaseVersionOverride` check once up front

File: docs/content/installation/gateway/kubernetes/values.txt (hunk 4 above)
Comment (ashleywang1) [reply docs/content/installation/gateway/kubernetes/values.txt]: aaahhhh i don't know what causes this it's so annoying! Reverting.

File: projects/gloo/cli/pkg/flagutils/install.go (hunk 6 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/flagutils/install.go]: Removing AddEnterpriseInstallFlags

File: docs/content/installation/gateway/kubernetes/values.txt (hunk 4 above)
Comment (kdorosh) [reply docs/content/installation/gateway/kubernetes/values.txt]: This happens when we run helm tests locally vs build release test assets. The charts get built with different values overrides and it manifests like this. We have an open issue for this somewhere, should make helm value generation part of the `make generated-code` step

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 8]
```diff
@@ -188,31 +190,35 @@ func (i *installer) printReleaseManifest(release *release.Release) error {
...
-	if err != nil {
-		return "", err
-	}
-	// Overrides
-	if version.EnterpriseTag != version.UndefinedVersion {
-		enterpriseTag = version.EnterpriseTag
+func getChartUri(chartOverride, releaseVersionOverride string, withUi, enterprise bool) (string, error) {
+
+	if chartOverride != "" && releaseVersionOverride != "" {
+		return "", errors.Errorf("you may only provide one override value, detected two: %s and %s", chartOverride, releaseVersionOverride)
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]:
```suggestion
		return "", errors.Errorf("you may not specify both a chart with -f and a release version with --release-version. Received: %s and %s", chartOverride, releaseVersionOverride)
```
Little wording nit, but I'm not sure that we've really established the idea of an "override" as a noun in users' heads. Better to be explicit here.

Also, can we change this error declaration into a function in an `errors.go` file, and get some test coverage of this path?

File: projects/gloo/cli/pkg/cmd/options/options.go [hunk 9]
```diff
@@ -51,6 +51,7 @@ type Install struct {
 	HelmChartOverride       string
 	HelmChartValueFileNames []string
 	HelmReleaseName         string
+	HelmReleaseVersion      string
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/options/options.go]: This isn't really a Helm specific release version. Can this just be ReleaseVersion?

File: projects/gloo/cli/pkg/flagutils/install.go [hunk 10]
```diff
@@ -13,13 +13,11 @@ func AddInstallFlags(set *pflag.FlagSet, install *options.Install) {
 	set.StringVarP(&install.HelmChartOverride, "file", "f", "", "Install Gloo from this Helm chart archive file rather than from a release")
 	set.StringSliceVarP(&install.HelmChartValueFileNames, "values", "", []string{}, "List of files with value overrides for the Gloo Helm chart, (e.g. --values file1,file2 or --values file1 --values file2)")
 	set.StringVar(&install.HelmReleaseName, "release-name", constants.GlooReleaseName, "helm release name")
+	set.StringVar(&install.HelmReleaseVersion, "release-version", "", "helm release version (e.g. 0.20.6, defaults to latest)")
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/flagutils/install.go]: same comment about usage; this isn't really Helm-specific. And since most people are going to be using open source, we might want to point to a different version for an example than 0.20.6. Also, is that going to need a "v" on it?

Comment (ashleywang1) [reply projects/gloo/cli/pkg/flagutils/install.go]: In the gloo helm charts, the strings are all formatted as:
```
GlooHelmRepoTemplate       = "https://storage.googleapis.com/solo-public-helm/charts/gloo-%s.tgz"
	GlooWithUiHelmRepoTemplate = "https://storage.googleapis.com/gloo-os-ui-helm/charts/gloo-os-with-ui-%s.tgz"
```
The actual files don't have a 'v' in them.

Comment (kdorosh) [reply projects/gloo/cli/pkg/flagutils/install.go]: Let's support both with and without the v

File: projects/gloo/cli/pkg/flagutils/install.go [hunk 11]
```diff
@@ -12,13 +12,11 @@ func AddInstallFlags(set *pflag.FlagSet, install *options.Install) {
 	set.StringVarP(&install.HelmChartOverride, "file", "f", "", "Install Gloo from this Helm chart archive file rather than from a release")
 	set.StringSliceVarP(&install.HelmChartValueFileNames, "values", "", []string{}, "List of files with value overrides for the Gloo Helm chart, (e.g. --values file1,file2 or --values file1 --values file2)")
 	set.StringVar(&install.HelmReleaseName, "release-name", constants.GlooReleaseName, "helm release name")
+	set.StringVar(&install.ReleaseVersion, "release-version", "", "release version to install (e.g. 0.20.6, defaults to latest)")
```
Comment (marcogschmidt) [inline projects/gloo/cli/pkg/flagutils/install.go]: Why don't we simply call this `version`?

CC: @kdorosh @GrahamGoudeau

Comment (kdorosh) [reply projects/gloo/cli/pkg/flagutils/install.go]: I agree, shorter is better

**Comment by ashleywang1** [comment]
/kick TestIngress failure:
Error: installing gloo in ingress mode: rendered manifests contain a resource that already exists. Unable to continue with install: existing resource conflict: kind: ClusterRole, namespace: , name: gloo-role-ingress
Error from server (AlreadyExists): namespaces "ingress-test-3394-1" already exists
/skip-changelog

File: pkg/version/linked.go [hunk 12]
```diff
@@ -8,3 +10,10 @@ var Version = UndefinedVersion
 func IsReleaseVersion() bool {
 	return Version != UndefinedVersion && Version != DevVersion
 }
+
+func StripV(version string) string {
+	if strings.HasPrefix(version, "v") {
+		return strings.TrimPrefix(version, "v")
```
[the remark this answers is not in the document]
Comment (ashleywang1) [reply pkg/version/linked.go]: yep, deleting the function

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 13]
```diff
@@ -188,31 +190,35 @@ func (i *installer) printReleaseManifest(release *release.Release) error {
...
+			if versionOverride != "" {
+				helmChartArchiveUri = fmt.Sprintf(constants.GlooHelmRepoTemplate, versionOverride)
+			} else {
+				glooOsVersion, err := getGlooVersionToInstall(chartOverride)
+				if err != nil {
+					return "", err
+				}
+				helmChartArchiveUri = fmt.Sprintf(constants.GlooHelmRepoTemplate, glooOsVersion)
+			}
+		}
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/cmd/install/installer.go]: nit, this is still very nested. we should be able to flatten this out and make it more readable.

either by:
```golang
if versionOverride != "" {
    helmChartArchiveUri = fmt.Sprintf(constants.GlooHelmRepoTemplate, versionOverride)
} else {
    glooOsVersion, err := getGlooVersionToInstall(chartOverride)
    if err != nil {
	return "", err
    }
    helmChartArchiveUri = fmt.Sprintf(constants.GlooHelmRepoTemplate, glooOsVersion)
    }
...
if enterprise {
    helmChartArchiveUri = chartUriWithVersion(GlooEHelmRepoTemplate, versionOverride, enterpriseVersion)
} else if withUi {
    helmChartArchiveUri = chartUriWithVersion(constants.GlooWithUiHelmRepoTemplate, versionOverride, enterpriseVersion)
		
```
or breaking some of it out into functions

File: changelog/validation.yaml [hunk 14]
```diff
@@ -0,0 +1 @@
+relaxSemverValidation: true
```
Comment (kdorosh) [inline changelog/validation.yaml]: 🎉

File: projects/gloo/cli/pkg/cmd/install/installer.go (hunk 13 above)
Comment (ashleywang1) [reply projects/gloo/cli/pkg/cmd/install/installer.go]: The reason I didn't break that part out is because this line:
```
glooOsVersion, err := getGlooVersionToInstall(chartOverride)
```
shouldn't get run if either enterprise or withUi is set to true. It'll error if the glooOsVersion is not a release version.

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 15]
```diff
@@ -236,14 +240,22 @@ func getChartUri(chartOverride string, withUi bool, enterprise bool) (string, er
 	return helmChartArchiveUri, nil
 }
 
-func getGlooVersion(chartOverride string) (string, error) {
+func getGlooVersionToInstall(chartOverride string) (string, error) {
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]: Couple thoughts:
* Should probably name this `getDefaultGlooVersion...` or something, since it just returns the linked version. However, backing up a level...
* I'm not even convinced this function should exist in its current form- we could move this validation earlier in the execution and then we don't even need this function, we can just directly use `version.Version`

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 16]
```diff
@@ -203,31 +207,31 @@ func (i *installer) printReleaseManifest(release *release.Release) error {
...
-	if err != nil {
-		return "", err
-	}
-	// Overrides
-	if version.EnterpriseTag != version.UndefinedVersion {
-		enterpriseTag = version.EnterpriseTag
+func getChartUri(chartOverride, versionOverride string, withUi, enterprise bool) (string, error) {
+
+	if chartOverride != "" && versionOverride != "" {
+		return "", errors.Errorf("you may not specify both a chart with -f and a release version with --version. Received: %s and %s", chartOverride, versionOverride)
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]: could move this error into an `errors.go` file so that in the test you don't have to duplicate this message

File: projects/gloo/cli/pkg/cmd/install/install_test.go [hunk 17]
```diff
@@ -25,13 +28,32 @@ var _ = Describe("Install", func() {
...
 		_, err := testutils.GlooctlOut(fmt.Sprintf("install gateway enterprise --file %s --dry-run %s", file, licenseKey))
 		Expect(err).NotTo(HaveOccurred())
 	})
 
+	It("shouldn't get errors for enterprise dry run without file", func() {
+		_, err := testutils.GlooctlOut(fmt.Sprintf("install gateway enterprise --dry-run %s", licenseKey))
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("shouldn't get errors when overriding enterprise version", func() {
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/install_test.go]: This whole test suite is fairly useless :/ not getting an error isn't the same as doing the right thing. Probably out of scope for the PR at this point to rewrite the whole file, I guess

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 18]
```diff
@@ -203,31 +207,31 @@ func (i *installer) printReleaseManifest(release *release.Release) error {
...
-func getChartUri(chartOverride string, withUi bool, enterprise bool) (string, error) {
-	var helmChartArchiveUri string
-	enterpriseTag, err := version.GetEnterpriseTag(true)
-	if err != nil {
-		return "", err
-	}
-	// Overrides
-	if version.EnterpriseTag != version.UndefinedVersion {
-		enterpriseTag = version.EnterpriseTag
+func getChartUri(chartOverride, versionOverride string, withUi, enterprise bool) (string, error) {
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]: Can we document the behavior of this function? eg, the general appearance of the values that the parameters can take (ie, chartOverride is a URI, versionOverride is some semver string without a leading "v", etc), and their interaction with each other

File: projects/gloo/cli/pkg/cmd/install/installer.go [hunk 19]
```diff
@@ -236,14 +240,22 @@ func getChartUri(chartOverride string, withUi bool, enterprise bool) (string, er
...
 }
 
-func getGlooVersion(chartOverride string) (string, error) {
+func getGlooVersionToInstall(chartOverride string) (string, error) {
 	if !version.IsReleaseVersion() && chartOverride == "" {
 		return "", errors.Errorf("you must provide a Gloo Helm chart URI via the 'file' option " +
 			"when running an unreleased version of glooctl")
 	}
 	return version.Version, nil
 }
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/cmd/install/installer.go]:
```suggestion

// Expects a format string containing a single string verb which will be replaced with `latestVersion`, or `releaseVersionOverride` instead if it is present
```

File: projects/gloo/cli/pkg/flagutils/install.go [hunk 20]
```diff
@@ -12,6 +12,7 @@ func AddInstallFlags(set *pflag.FlagSet, install *options.Install) {
 	set.StringVarP(&install.HelmChartOverride, "file", "f", "", "Install Gloo from this Helm chart archive file rather than from a release")
 	set.StringSliceVarP(&install.HelmChartValueFileNames, "values", "", []string{}, "List of files with value overrides for the Gloo Helm chart, (e.g. --values file1,file2 or --values file1 --values file2)")
 	set.StringVar(&install.HelmReleaseName, "release-name", constants.GlooReleaseName, "helm release name")
+	set.StringVar(&install.Version, "version", "", "version to install (e.g. 1.2.0, defaults to latest)")
```
Comment (GrahamGoudeau) [inline projects/gloo/cli/pkg/flagutils/install.go]:
```suggestion
	set.StringVar(&install.Version, "version", "", "version of Gloo to install (e.g. 1.2.0, defaults to latest)")
```
to disambiguate an invocation like `glooctl install knative --version 1.2.0`

File: pkg/version/enterprise.go [hunk 21]
```diff
@@ -7,14 +7,12 @@ import (
...
 
-var EnterpriseTag = UndefinedVersion
-
 const EnterpriseHelmRepoIndex = "https://storage.googleapis.com/gloo-ee-helm/index.yaml"
 const GlooEE = "gloo-ee"
 
 // The version of GlooE installed by the CLI.
 // Calculated from the latest gloo-ee version in the helm repo index
-func GetEnterpriseTag(stableOnly bool) (string, error) {
+func GetLatestEnterpriseVersion(stableOnly bool) (string, error) {
```
Comment (GrahamGoudeau) [inline pkg/version/enterprise.go]: why the change in noun from Tag to Version? Makes at least one usage inconsistent, eg

`enterpriseTag, err := version.GetLatestEnterpriseVersion(false)`

File: pkg/utils/setuputils/main_setup.go [hunk 22]
```diff
@@ -24,6 +24,7 @@ import (
 
 type SetupOpts struct {
 	LoggerName  string
```
Comment (GrahamGoudeau) [inline pkg/utils/setuputils/main_setup.go]:
```suggestion
	LoggerName  string
	
	// logged as the version of Gloo currently executing
```

File: projects/gloo/cli/pkg/flagutils/install.go (hunk 20 above)
Comment (GrahamGoudeau) [reply projects/gloo/cli/pkg/flagutils/install.go]: Follow up question: Is `glooctl install knative --version ...` still too ambiguous?

File: projects/gloo/cli/pkg/cmd/install/installer.go (hunk 15 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/install/installer.go]: This validation has to happen last (after all other checks), so can't move it earlier. Happy to rename though

File: projects/gloo/cli/pkg/cmd/install/installer.go (hunk 18 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/install/installer.go]: I rewrote the whole thing, i think it should be clearer now

File: pkg/version/enterprise.go (hunk 21 above)
Comment (kdorosh) [reply pkg/version/enterprise.go]: Updated to be consistent

File: pkg/utils/setuputils/main_setup.go (hunk 22 above)
Comment (kdorosh) [reply pkg/utils/setuputils/main_setup.go]: https://github.com/solo-io/gloo/pull/1968/commits/be56a242619efd0b0d2fcff4a0c322b788e4e4ae

File: pkg/version/enterprise.go (hunk 21 above)
Comment (kdorosh) [reply pkg/version/enterprise.go]: https://github.com/solo-io/gloo/pull/1968/commits/f77d748e6c0d1eba0b3b87e4b97f29381a61b2bf

File: projects/gloo/cli/pkg/cmd/install/installer.go (hunk 19 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/install/installer.go]: removed the function instead here https://github.com/solo-io/gloo/pull/1968/commits/e4e874444c399d2d0896e99e99c3360d40dfaea5

File: projects/gloo/cli/pkg/cmd/install/install_test.go (hunk 17 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/cmd/install/install_test.go]: I actually found these tests useful while rewriting the `getChart` function, would like to leave them around

File: projects/gloo/cli/pkg/flagutils/install.go (hunk 20 above)
Comment (kdorosh) [reply projects/gloo/cli/pkg/flagutils/install.go]: I think it's fine, @marcogschmidt thoughts?

Comment (kdorosh) [reply projects/gloo/cli/pkg/flagutils/install.go]: @ilackarms thoughts?

Comment (marcogschmidt) [reply projects/gloo/cli/pkg/flagutils/install.go]: It is fine with me, as long as the doc for the flag is clear.
